File: src/main/java/io/github/jho951/assignment/job/worker/WorkerClientException.java
package io.github.jho951.assignment.job.worker;

import io.github.jho951.assignment.job.domain.JobFailureCode;

public class WorkerClientException extends RuntimeException {

    private final JobFailureCode failureCode;
    private final boolean retryable;

    public WorkerClientException(JobFailureCode failureCode, boolean retryable, String message, Throwable cause) {
        super(message, cause);
        this.failureCode = failureCode;
        this.retryable = retryable;
    }

    public WorkerClientException(JobFailureCode failureCode, boolean retryable, String message) {
        super(message);
        this.failureCode = failureCode;
        this.retryable = retryable;
    }

    public JobFailureCode getFailureCode() {
        return failureCode;
    }

    public boolean isRetryable() {
        return retryable;
    }
}

File: src/main/kotlin/io/github/jho951/assignment/job/worker/WorkerClientException.kt
package io.github.jho951.assignment.job.worker

import io.github.jho951.assignment.job.domain.JobFailureCode

open class WorkerClientException(
    val failureCode: JobFailureCode,
    private val retryable: Boolean,
    message: String,
    cause: Throwable? = null
) : RuntimeException(message, cause) {

    constructor(
        failureCode: JobFailureCode,
        retryable: Boolean,
        message: String
    ) : this(failureCode, retryable, message, null)

    fun isRetryable(): Boolean = retryable
}
